libraries/type-length-value/src/state.rs

```diff
@@ -861,19 +1103,71 @@ mod test {
+
+        // realloc first entry to larger, all 0
+        let data = state
+            .realloc_unique::<TestValue>(TLV_SIZE + EXTRA_SPACE)
+            .unwrap();
+        assert_eq!(data, [0; TLV_SIZE + EXTRA_SPACE]);
+        let value = state.get_first_value::<TestNonZeroDefault>().unwrap();
+        assert_eq!(*value, TestNonZeroDefault::default());
+
+        // realloc to smaller, still all 0
+        let data = state.realloc_unique::<TestValue>(SMALL_SIZE).unwrap();
+        assert_eq!(data, [0; SMALL_SIZE]);
+        let value = state.get_first_value::<TestNonZeroDefault>().unwrap();
+        assert_eq!(*value, TestNonZeroDefault::default());
+        let (_, end_index) = get_discriminators_and_end_index(&buffer).unwrap();
+        assert_eq!(
+            &buffer[end_index..ACCOUNT_SIZE],
+            [0; TLV_SIZE + EXTRA_SPACE - SMALL_SIZE]
+        );
+
+        // unpack again since we dropped the last `state`
+        let mut state = TlvStateMut::unpack(&mut buffer).unwrap();
+        // realloc too much, fails
+        assert_eq!(
+            state
+                .realloc_unique::<TestValue>(TLV_SIZE + EXTRA_SPACE + 1)
+                .unwrap_err(),
+            ProgramError::InvalidAccountData,
+        );
+    }
+
+    #[test]
+    fn realloc_with_repeating_entries() {
+        const TLV_SIZE: usize = 10;
+        const EXTRA_SPACE: usize = 5;
+        const SMALL_SIZE: usize = 2;
+        const ACCOUNT_SIZE: usize = get_base_len()
+            + TLV_SIZE
+            + EXTRA_SPACE
+            + get_base_len()
+            + TLV_SIZE
+            + get_base_len()
+            + size_of::<TestNonZeroDefault>();
+        let mut buffer = vec![0; ACCOUNT_SIZE];
+        let mut state = TlvStateMut::unpack(&mut buffer).unwrap();
+
+        // alloc both types, two for the first type and one for the second
+        let _ = state.alloc_allow_repeating::<TestValue>(TLV_SIZE).unwrap();
+        let _ = state.alloc_allow_repeating::<TestValue>(TLV_SIZE).unwrap();
         let _ = state.init_value::<TestNonZeroDefault>().unwrap();
 
         // realloc first entry to larger, all 0
-        let data = state.realloc::<TestValue>(TLV_SIZE + EXTRA_SPACE).unwrap();
+        let data = state
+            .realloc::<TestValue>(TLV_SIZE + EXTRA_SPACE, 0)
+            .unwrap();
         assert_eq!(data, [0; TLV_SIZE + EXTRA_SPACE]);
-        let value = state.get_value::<TestNonZeroDefault>().unwrap();
+        let value = state.get_first_value::<TestNonZeroDefault>().unwrap();
         assert_eq!(*value, TestNonZeroDefault::default());
 
         // realloc to smaller, still all 0
-        let data = state.realloc::<TestValue>(SMALL_SIZE).unwrap();
+        let data = state.realloc::<TestValue>(SMALL_SIZE, 0).unwrap();
         assert_eq!(data, [0; SMALL_SIZE]);
-        let value = state.get_value::<TestNonZeroDefault>().unwrap();
+        let value = state.get_first_value::<TestNonZeroDefault>().unwrap();
```

Comment:
You're checking `TestNonZeroDefault`, but to be sure, can you also check that the other `TestValue` entry is all zeroes after each realloc?